File: LeetCode/Medium/0049-group-anagrams/0049-group-anagrams.js
/*
P:
    input: array of strings (strs)
    output 2D matrix of strings (grouped anagrams)]
    task: return the grouped anagrams together in arrays within the array
 
E:
    - the order of the grouped anagrams can be returned in any order
    - the input can have an empty string: return empty array if so
    - the input won't be empty: guranteed at least one string (even if empty)

D:
    strings, arrays, hash tables

A:
    1: initiate an empty hash map (to store anagrams)
    2. iterate through input array with for loop O(n)
        2a. store current word in a variable
        2b. sort the current word alphabetically O(n log n)
        2b. check to see if this word has been seen already within hash
            2ba. if it has, just add the word to the already existing key ("word.sorted": [word1, word2])
            2bb. else, create the key-value pair and add it
    3. iterate through hash table
        3a. take all values and push it into output array
    4. return final output
            
*/

/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    const hash = {};
    const outputArr = [];

    for (let i = 0; i < strs.length; i++) {
        const currentWord = strs[i];
        const wordSorted = currentWord.split('').sort().join('');

        if (hash.hasOwnProperty(wordSorted)) {
            hash[wordSorted].push(currentWord);
        } else {
            hash[wordSorted] = [currentWord];
        }
    }
    
    for (const key in hash) {
        outputArr.push(hash[key]);
    }
    return outputArr;
};

File: LeetCode/Medium/0049-group-anagrams/0049-group-anagrams.py
"""
P:
    input: list of strs (strs)
    output: 2-D array (matrix)
    output: group all the anagrams together and return them (in order)
        - an anagram is a word/phrase formed by rearranging the letters of a different word or phrase, using all the original letters exactly once.

E:
    - input will never be empty, at least have 1 str in it
    - strs can be empty or have 1 letter, in that case just return it in its own list

D:
    - strings, lists, matrixes 
"""
from collections import defaultdict
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        sorted_dict = defaultdict(list) # auto-assigns an empty list to a new key in the dict

        for word in strs:
            sorted_str = "".join(sorted(word)) # sort each string's characters alphabetically to create a hashable key

            sorted_dict[sorted_str].append(word) # group the original string under its sorted key

        return list(sorted_dict.values()) # returns the substrings (dict values) in one list together